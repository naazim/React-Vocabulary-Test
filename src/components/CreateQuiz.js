import React from 'react';
import Options from './Options';
import AddOption from './AddOption';
import Progress from './Progress';

export default class CreateQuiz extends React.Component {
    constructor(props) {
        super(props);
    };

    state = {
        wordList: [],
    };

    handleDeleteOption = (optionToRemove) => {
        this.setState((prevState) => ({
            wordList: prevState.wordList.filter((word) => optionToRemove !== prevState.wordList.indexOf(word))
        }));
    };

    handleDeleteOptions = () => {
        this.setState(() => ({ wordList: [] }));
    };

    handleClearSelectedOption = () => {
        this.setState(() => ({ selectedOption: undefined }));
    };

    handleTest = () => {
        const randomNum = Math.floor(Math.random() * this.state.wordList.length);
        const option = this.state.wordList[randomNum];
        this.props.handleModule('PerformQuiz');
    };

    handleAddOption = (option, option2) => {
        if (!option) {
            return 'Enter valid native word';
        }
        if (!option2) {
            return 'Enter valid foriegn word';
        }
        else if (this.state.wordList.indexOf(option) > -1 || this.state.wordList.indexOf(option2) > -1) {
            return 'This word already exists';
        }

        this.setState((prevState) => ({
            wordList: [...prevState.wordList, { 'native': option, 'foreign': option2, 'userAnswer': undefined }]
        }));

    };

    componentDidMount() {
        try {
            const json = localStorage.getItem('wordList');
            const wordList = JSON.parse(json);

            if (wordList) {
                this.setState(() => ({ wordList }));
            }
        } catch (e) {
            // Do nothing at all
        }
    }
    componentDidUpdate(prevProps, prevState) {
        if (prevState.wordList.length !== this.state.wordList.length) {
            const json = JSON.stringify(this.state.wordList);
            localStorage.setItem('wordList', json);
        };
    }
    componentWillUnmount() {
        console.log('componentWillUnmount');
    }

    render() {
        const {wordList} = this.state;
        return (
            <div className="wrapper">
                <Progress percentage={wordList.length} />
                <div className="widget">
                    <Options
                        options={wordList}
                        handleDeleteOptions={this.handleDeleteOptions}
                        handleDeleteOption={this.handleDeleteOption}
                    />
                    <AddOption
                        handleAddOption={this.handleAddOption}
                    />
                </div>
                <input
                    type="button"
                    onClick={this.handleTest}
                    className="button button--big"
                    disabled={wordList.length < 2}
                    value="Take Test"
                />
            </div>
        );
    }
};