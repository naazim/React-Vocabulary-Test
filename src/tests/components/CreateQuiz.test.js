import React from 'react';
import {shallow} from 'enzyme';
import CreateQuiz, {percentage} from '../../components/CreateQuiz';

describe('Create Quiz', () => {
    const handleModule = jest.fn();
    let wordList = [{"native":"Hello","foreign":"Hallo"},{"native":"Good","foreign":"Gut"}];
    let wrapper = shallow(<CreateQuiz wordList={wordList} handleModule={handleModule} />);
    
    test('should create the quiz correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });
    
    test('should show the percentage correctly', () => {
        expect(wrapper.find('Progress').props().percentage).toBe(10);
    });
    
    test('should disable button if word count is less than 20', () => {
        expect(wrapper.find('.button--big').props().disabled).toBe(true);
    });
    
    test('should enable button if word count is more than 20', () => {
        wordList = new Array(20);
        wrapper = shallow(<CreateQuiz wordList={wordList} handleModule={handleModule} />);
        expect(wrapper.find('.button--big').props().disabled).toBe(false);
    });
});


