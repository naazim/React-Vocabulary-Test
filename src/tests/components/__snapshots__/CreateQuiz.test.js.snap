// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Quiz should create the quiz correctly 1`] = `
<div
  className="wrapper"
>
  <Progress
    percentage={10}
  />
  <div
    className="widget"
  >
    <Options
      options={
        Array [
          Object {
            "foreign": "Hallo",
            "native": "Hello",
          },
          Object {
            "foreign": "Gut",
            "native": "Good",
          },
        ]
      }
    />
    <AddWord
      handleAddWord={[Function]}
    />
  </div>
  <input
    className="button button--big"
    disabled={true}
    onClick={[Function]}
    type="button"
    value="Take Test"
  />
</div>
`;
